src/Microsoft.IdentityModel.Tokens.Saml/Saml2/Saml2Assertion.cs

```diff
@@ -51,101 +59,179 @@ public Saml2Assertion(Saml2NameIdentifier issuer)
             Id = new Saml2Id();
             IssueInstant = DateTime.UtcNow;
             Issuer = issuer;
-            Statements = new List<Saml2Statement>();
+            _statements = new List<Saml2Statement>();
         }
 
         /// <summary>
         /// Gets or sets the <see cref="Signature"/> on the Assertion.
         /// </summary>
+        /// <exception cref="Saml2SecurityTokenEncryptedAssertionException"> If this assertion is encrypted.</exception>
         public Signature Signature
         {
-            get;
-            set;
+            get
+            {
+                if (Encrypted)
+                    throw LogExceptionMessage(new Saml2SecurityTokenEncryptedAssertionException(FormatInvariant(LogMessages.IDX13608, nameof(Signature))));
+
+                return _signature;
+            }
+            set => _signature = value;
         }
 
         /// <summary>
         /// Gets or sets additional information related to the assertion that assists processing in certain
         /// situations but which may be ignored by applications that do not understand the 
         /// advice or do not wish to make use of it. [Saml2Core, 2.3.3]
         /// </summary>
+        /// <exception cref="Saml2SecurityTokenEncryptedAssertionException"> If this assertion is encrypted.</exception>
         public Saml2Advice Advice
         {
-            get;
-            set;
+            get
+            {
+                if (Encrypted)
```

Comment:
We could just return null here, and not throw.
The use case is a user created a SAML token with just encrypted assertion, that would be the only value that is not null.

This would allow us to remove the type Saml2SecurityTokenEncryptedAssertionException.